rfcs/0024-hard-forks.md

```diff
@@ -0,0 +1,455 @@
+## Summary
+[summary]: #summary
+
+This RFC describes a system for planning and enacting hard forks.
+
+## Motivation
+
+[motivation]: #motivation
+
+Whenever we make a change to the protocol that changes the rules of consensus or
+the state transition system, the network must switch to the new system
+atomically, and on all nodes simultaneously. Changes of this sort include for
+example modifying the parameters of the parallel scan, making them dynamic,
+adding smart contracts, adding shielded transactions/accounts, modifying or
+replacing Codaborous, and many others.
+
+Changes that do *not* require hard forks include changes to the network
+protocol but not the underlying state machine, and anything that doesn't affect
+other nodes. We may want to do a hard fork anyway in some of these situations to
+force users to upgrade, especially if the changes are not backwards compatible
+or are very important security improvements.
+
+If a change of the first sort were not implemented atomically and at a
+coordinated time, there would an unintentional fork: nodes supporting the new
+version would generate and accept blocks with the new specification, and nodes
+not supporting it would generate and accept blocks with the old one. To prevent
+this we plan the fork in advance, and make sure out of date clients stop working
+if they are not updated in time.
+
+## Detailed design
+
+[detailed-design]: #detailed-design
+
+I'll call a period of time where one state machine/consensus system is in force
+an *era*, to make clear the distinction between them and protocol versions and
+epochs. Era zero is the one that is in force at the launch of a network, era one
+follows after the first hard fork, and so on. Transitions between eras occur at
+a pre-specified time - all blocks with a global slot that begins at or after the
+specified era start time and before the era end time will follow the consensus
+rules of that era. An *era transition block* is the first block of a new era. If
+the parallel scan state is changing or for some other reason we need it, there
+may be a special intermediate state. See below for details.
+
+I'll separate the considerations in this RFC into "human" considerations and
+"technical" ones. Human considerations refers to how the process should work for
+us as we plan, test, and execute the forks. Technical considerations refers to
+how the forks will be implemented in the client.
+
+### Human considerations
```

Comment:
If we're going to include the logistics we may want to involve more people (BizDev, Community, ...) that would be impacted by any advance planning